src/hooks/useDebounce.js

```diff
@@ -0,0 +1,31 @@
+import { useCallback } from 'react';
+
+export const useDebounce = (func, waitAmount, immediate) => {
+  const debounceIt = () => {
+    // Closure timeout ref for all subsequent calls
+    let timeout;
+
+    // Return a func wrapper with the timeout var ref closure
+    return (...args) => {
+      const doItLater = () => {
+        timeout = null;
+        if (!immediate) func(...args);
+        return null;
+      };
+
+      // If the immediate flag is true and there is no timeout
+      // already, just execute.
+      const doItNow = immediate && !timeout;
```

Comment:
Assuming `!timeout` means clearing `timeout` will be redundant, can you not do:

```suggestion
      if (immediate && !timeout) return func(...args);
```

and remove the `if (doItNow)` conditional at the end? This would possibly remove the need to handle double executions that could occur otherwise?

Reply:
clearing timeout is redundant if there isn't one already, yeah, but you still need to set one after you execute the func so it debounces any further invocations

Doing it this way you would never set a timer, though or am I misunderstanding? So what you suggested would be equivalent to `debounce = (func, wait, imm) => func`..?

Reply:
Wouldn't you still set the timer is `immediate` was falsey? Also, if `immediate` is truthy doesn't it stop `doItLater` from executing anyway?

Reply:
Ah now I see: 

Yes, you would set the timer if immediate was false and it would work fine. But,if it was true, then the timer would never start and you would never debounce.

With this implementation, when immediate is true, you always execute on the first invocation, and then start the timer which will block any further invocations from executing the func, but the func is not executed the second time.